File: src/atoms/time-tracking.ts
import { atom } from "jotai";
import { atomWithStorage } from "jotai/utils";
import { message } from "antd";
import { nanoid } from "nanoid";
import { t } from "@lingui/core/macro";
import dayjs from "dayjs";
import isEqual from "lodash/isEqual";
import orderBy from "lodash/orderBy";
import keyBy from "lodash/keyBy";
import map from "lodash/map";
import reject from "lodash/reject";
import { invoke } from "@tauri-apps/api/core";

import { organizationIdAtom } from "./organization";

// Tags
export const tagsAtom = atom<any[]>([]);
export const setTagsAtom = atom(null, async (get, set) => {
  const organizationId = get(organizationIdAtom);
  try {
    const response = await invoke<any[]>("get_tags", { organizationId });
    set(tagsAtom, response);
  } catch (error) {
    console.error("Failed to fetch tags:", error);
    message.error(t`Failed to fetch tags`);
  }
});

// Tag
export const tagIdAtom = atom<string | null>(null);
export const tagAtom = atom(
  async (get) => {
    const tagId = get(tagIdAtom);
    if (!tagId) return null;

    try {
      const tag = await invoke<any>("get_tag", { tagId });
      return tag;
    } catch (error) {
      console.error("Failed to fetch tag:", error);
      return null;
    }
  },
  async (get, set, newValues: any) => {
    const tagId = get(tagIdAtom);

    try {
      if (!tagId) {
        // Insert
        const tagData = {
          ...newValues,
          id: nanoid(),
          organizationId: get(organizationIdAtom),
        };

        const createdTag = await invoke<any>("create_tag", {
          tag: tagData,
        });
        set(tagIdAtom, createdTag.id);
        message.success(t`Tag created`);

        // Update the tags list
        const tags: any = get(tagsAtom);
        set(tagsAtom, orderBy([...tags, createdTag], "name", "asc"));
      } else {
        // Update
        const updatedTag = await invoke<any>("update_tag", {
          tagId,
          updates: newValues,
        });
        message.success(t`Tag updated successfully`);

        // Update the tags list
        const tags: any = get(tagsAtom);
        const mergedTags: any = keyBy([...tags, updatedTag], "id");
        set(tagsAtom, orderBy(map(mergedTags), "name", "asc"));
      }
    } catch (error) {
      console.error("Tag operation failed:", error);
      if (!tagId) {
        message.error(t`Tag creation failed`);
      } else {
        message.error(t`Tag update failed`);
      }
    }
  }
);

// Delete tag
export const deleteTagAtom = atom(null, async (get, set, tagId: string) => {
  try {
    const success = await invoke<boolean>("delete_tag", { tagId });

    if (success) {
      // Remove tag from the list
      const tags: any = reject(get(tagsAtom), (obj: any) => isEqual(obj.id, tagId));
      set(tagsAtom, tags);
      message.success(t`Tag deleted`);
    } else {
      message.error(t`Tag deletion failed`);
    }
  } catch (error) {
    console.error("Failed to delete tag:", error);
    message.error(t`Tag deletion failed`);
  }
});

// Time Entries
export const timeEntriesAtom = atom<any[]>([]);
export const setTimeEntriesAtom = atom(null, async (get, set) => {
  const organizationId = get(organizationIdAtom);
  try {
    const response = await invoke<any[]>("get_time_entries", { organizationId });
    set(timeEntriesAtom, response);
  } catch (error) {
    console.error("Failed to fetch time entries:", error);
    message.error(t`Failed to fetch time entries`);
  }
});

// Update time entry locally without refetching
export const updateTimeEntryLocallyAtom = atom(null, (get, set, { id, updates }: { id: string, updates: any }) => {
  const entries = get(timeEntriesAtom);
  const updatedEntries = entries.map(entry => 
    entry.id === id ? { ...entry, ...updates } : entry
  );
  set(timeEntriesAtom, updatedEntries);
});

// Direct update time entry without triggering fetches
export const updateTimeEntryDirectlyAtom = atom(null, async (get, set, { id, updates }: { id: string, updates: any }) => {
  const entry = get(timeEntriesAtom).find(e => e.id === id);
  
  if (!entry) return;
  
  try {
    // Merge updates with existing entry
    const updatedEntry = { ...entry, ...updates };
    
    // Prepare the update data for database
    const updateData = {
      ...updatedEntry,
      // Convert dayjs back to timestamps if needed
      startTime: updatedEntry.startTime instanceof dayjs ? updatedEntry.startTime.valueOf() : updatedEntry.startTime,
      endTime: updatedEntry.endTime instanceof dayjs ? updatedEntry.endTime.valueOf() : updatedEntry.endTime,
      tags: Array.isArray(updatedEntry.tags) ? JSON.stringify(updatedEntry.tags) : updatedEntry.tags,
    };
    
    // Update in database
    await invoke("update_time_entry", { timeEntryId: id, updates: updateData });
    
    // Update local state immediately
    set(updateTimeEntryLocallyAtom, { id, updates });
    
  } catch (error) {
    console.error("Failed to update time entry:", error);
    message.error(t`Failed to update time entry`);
  }
});

// Time Entry
export const timeEntryIdAtom = atom<string | null>(null);
export const timeEntryAtom = atom(
  async (get) => {
    const timeEntryId = get(timeEntryIdAtom);
    if (!timeEntryId) return null;

    try {
      const timeEntry = await invoke<any>("get_time_entry", { timeEntryId });
      if (!timeEntry) return null;

      return {
        ...timeEntry,
        startTime: timeEntry.startTime ? dayjs(timeEntry.startTime) : null,
        endTime: timeEntry.endTime ? dayjs(timeEntry.endTime) : null,
        tags: timeEntry.tags ? JSON.parse(timeEntry.tags) : [],
      };
    } catch (error) {
      console.error("Failed to fetch time entry:", error);
      return null;
    }
  },
  async (get, set, newValues: any) => {
    const timeEntryId = get(timeEntryIdAtom);

    try {
      if (!timeEntryId) {
        // Insert
        const timeEntryData = {
          ...newValues,
          id: nanoid(),
          organizationId: get(organizationIdAtom),
          // Convert dayjs objects to unix timestamps
          startTime: newValues.startTime?.valueOf ? newValues.startTime.valueOf() : newValues.startTime,
          endTime: newValues.endTime?.valueOf ? newValues.endTime.valueOf() : newValues.endTime,
          // Convert tags array to JSON string
          tags: Array.isArray(newValues.tags) ? JSON.stringify(newValues.tags) : newValues.tags,
          // Convert boolean to integer
          isBillable: typeof newValues.isBillable === "boolean" ? (newValues.isBillable ? 1 : 0) : newValues.isBillable,
        };

        const createdTimeEntry = await invoke<any>("create_time_entry", {
          timeEntry: timeEntryData,
        });
        set(timeEntryIdAtom, createdTimeEntry.id);
        message.success(t`Time entry created`);

        // Update the time entries list
        const timeEntries: any = get(timeEntriesAtom);
        set(timeEntriesAtom, [createdTimeEntry, ...timeEntries]);
        
        return createdTimeEntry;
      } else {
        // Update
        const updateData = {
          ...newValues,
          // Convert dayjs objects to unix timestamps
          startTime: newValues.startTime?.valueOf ? newValues.startTime.valueOf() : newValues.startTime,
          endTime: newValues.endTime?.valueOf ? newValues.endTime.valueOf() : newValues.endTime,
          // Convert tags array to JSON string
          tags: Array.isArray(newValues.tags) ? JSON.stringify(newValues.tags) : newValues.tags,
          // Convert boolean to integer
          isBillable: typeof newValues.isBillable === "boolean" ? (newValues.isBillable ? 1 : 0) : newValues.isBillable,
        };

        const updatedTimeEntry = await invoke<any>("update_time_entry", {
          timeEntryId,
          updates: updateData,
        });
        message.success(t`Time entry updated successfully`);

        // Update the time entries list
        const timeEntries: any = get(timeEntriesAtom);
        const mergedTimeEntries: any = keyBy([...timeEntries, updatedTimeEntry], "id");
        set(timeEntriesAtom, orderBy(map(mergedTimeEntries), "startTime", "desc"));
        
        return updatedTimeEntry;
      }
    } catch (error) {
      console.error("Time entry operation failed:", error);
      if (!timeEntryId) {
        message.error(t`Time entry creation failed`);
      } else {
        message.error(t`Time entry update failed`);
      }
    }
  }
);

// Delete time entry
export const deleteTimeEntryAtom = atom(null, async (get, set, timeEntryId: string) => {
  try {
    const success = await invoke<boolean>("delete_time_entry", { timeEntryId });

    if (success) {
      // Remove time entry from the list
      const timeEntries: any = reject(get(timeEntriesAtom), (obj: any) => isEqual(obj.id, timeEntryId));
      set(timeEntriesAtom, timeEntries);
      message.success(t`Time entry deleted`);
    } else {
      message.error(t`Time entry deletion failed`);
    }
  } catch (error) {
    console.error("Failed to delete time entry:", error);
    message.error(t`Time entry deletion failed`);
  }
});

// Running timer atom for tracking active timer (persisted in localStorage)
export const runningTimerAtom = atomWithStorage<string | null>("runningTimer", null);

File: src-tauri/src/db/time_tracking.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

use super::Database;

// Tags
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Tag {
    pub id: String,
    #[serde(rename = "organizationId")]
    #[sqlx(rename = "organizationId")]
    pub organization_id: String,
    pub name: String,
    pub color: String,
    #[serde(rename = "createdAt")]
    #[sqlx(rename = "createdAt")]
    pub created_at: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CreateTagRequest {
    pub id: String,
    #[serde(rename = "organizationId")]
    pub organization_id: String,
    pub name: String,
    pub color: String,
}

#[derive(Debug, Deserialize)]
pub struct UpdateTagRequest {
    pub name: Option<String>,
    pub color: Option<String>,
}

// Time Entries
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct TimeEntry {
    pub id: String,
    #[serde(rename = "organizationId")]
    #[sqlx(rename = "organizationId")]
    pub organization_id: String,
    #[serde(rename = "clientId")]
    #[sqlx(rename = "clientId")]
    pub client_id: Option<String>,
    pub description: Option<String>,
    #[serde(rename = "startTime")]
    #[sqlx(rename = "startTime")]
    pub start_time: i64,
    #[serde(rename = "endTime")]
    #[sqlx(rename = "endTime")]
    pub end_time: Option<i64>,
    pub duration: i64,
    pub tags: Option<String>, // JSON array of tag names
    #[serde(rename = "isBillable")]
    #[sqlx(rename = "isBillable")]
    pub is_billable: i64,
    #[serde(rename = "hourlyRate")]
    #[sqlx(rename = "hourlyRate")]
    pub hourly_rate: Option<f64>,
    #[serde(rename = "createdAt")]
    #[sqlx(rename = "createdAt")]
    pub created_at: Option<String>,
    // Only for joined queries
    #[serde(rename = "clientName")]
    #[sqlx(rename = "clientName")]
    pub client_name: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CreateTimeEntryRequest {
    pub id: String,
    #[serde(rename = "organizationId")]
    pub organization_id: String,
    #[serde(rename = "clientId")]
    pub client_id: Option<String>,
    pub description: Option<String>,
    #[serde(rename = "startTime")]
    pub start_time: i64,
    #[serde(rename = "endTime")]
    pub end_time: Option<i64>,
    pub duration: i64,
    pub tags: Option<String>, // JSON array of tag names
    #[serde(rename = "isBillable")]
    pub is_billable: i64,
    #[serde(rename = "hourlyRate")]
    pub hourly_rate: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateTimeEntryRequest {
    #[serde(rename = "clientId")]
    pub client_id: Option<String>,
    pub description: Option<String>,
    #[serde(rename = "startTime")]
    pub start_time: Option<i64>,
    #[serde(rename = "endTime")]
    pub end_time: Option<i64>,
    pub duration: Option<i64>,
    pub tags: Option<String>, // JSON array of tag names
    #[serde(rename = "isBillable")]
    pub is_billable: Option<i64>,
    #[serde(rename = "hourlyRate")]
    pub hourly_rate: Option<f64>,
}

impl Database {
    // Tag methods
    pub async fn get_tags(&self, organization_id: &str) -> Result<Vec<Tag>, sqlx::Error> {
        sqlx::query_as::<_, Tag>("SELECT * FROM tags WHERE organizationId = ? ORDER BY name")
            .bind(organization_id)
            .fetch_all(&self.pool)
            .await
    }

    pub async fn get_tag(&self, tag_id: &str) -> Result<Option<Tag>, sqlx::Error> {
        sqlx::query_as::<_, Tag>("SELECT * FROM tags WHERE id = ?")
            .bind(tag_id)
            .fetch_optional(&self.pool)
            .await
    }

    pub async fn create_tag(&self, tag: CreateTagRequest) -> Result<Tag, sqlx::Error> {
        sqlx::query(
            "INSERT INTO tags (id, organizationId, name, color) VALUES (?, ?, ?, ?)"
        )
        .bind(&tag.id)
        .bind(&tag.organization_id)
        .bind(&tag.name)
        .bind(&tag.color)
        .execute(&self.pool)
        .await?;

        self.get_tag(&tag.id).await?
            .ok_or_else(|| sqlx::Error::RowNotFound)
    }

    pub async fn update_tag(&self, tag_id: &str, updates: UpdateTagRequest) -> Result<Tag, sqlx::Error> {
        sqlx::query(
            "UPDATE tags SET name = COALESCE(?, name), color = COALESCE(?, color) WHERE id = ?"
        )
        .bind(&updates.name)
        .bind(&updates.color)
        .bind(tag_id)
        .execute(&self.pool)
        .await?;

        self.get_tag(tag_id).await?
            .ok_or_else(|| sqlx::Error::RowNotFound)
    }

    pub async fn delete_tag(&self, tag_id: &str) -> Result<bool, sqlx::Error> {
        let result = sqlx::query("DELETE FROM tags WHERE id = ?")
            .bind(tag_id)
            .execute(&self.pool)
            .await?;

        Ok(result.rows_affected() > 0)
    }

    // Time Entry methods
    pub async fn get_time_entries(&self, organization_id: &str) -> Result<Vec<TimeEntry>, sqlx::Error> {
        sqlx::query_as::<_, TimeEntry>(
            "SELECT t.*, c.name as clientName 
             FROM timeEntries t 
             LEFT JOIN clients c ON t.clientId = c.id 
             WHERE t.organizationId = ? 
             ORDER BY t.startTime DESC"
        )
        .bind(organization_id)
        .fetch_all(&self.pool)
        .await
    }

    pub async fn get_time_entry(&self, time_entry_id: &str) -> Result<Option<TimeEntry>, sqlx::Error> {
        sqlx::query_as::<_, TimeEntry>(
            "SELECT t.*, c.name as clientName 
             FROM timeEntries t 
             LEFT JOIN clients c ON t.clientId = c.id 
             WHERE t.id = ?"
        )
        .bind(time_entry_id)
        .fetch_optional(&self.pool)
        .await
    }

    pub async fn create_time_entry(&self, time_entry: CreateTimeEntryRequest) -> Result<TimeEntry, sqlx::Error> {
        sqlx::query(
            "INSERT INTO timeEntries (id, organizationId, clientId, description, startTime, endTime, duration, tags, isBillable, hourlyRate) 
             VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
        )
        .bind(&time_entry.id)
        .bind(&time_entry.organization_id)
        .bind(&time_entry.client_id)
        .bind(&time_entry.description)
        .bind(&time_entry.start_time)
        .bind(&time_entry.end_time)
        .bind(&time_entry.duration)
        .bind(&time_entry.tags)
        .bind(&time_entry.is_billable)
        .bind(&time_entry.hourly_rate)
        .execute(&self.pool)
        .await?;

        self.get_time_entry(&time_entry.id).await?
            .ok_or_else(|| sqlx::Error::RowNotFound)
    }

    pub async fn update_time_entry(&self, time_entry_id: &str, updates: UpdateTimeEntryRequest) -> Result<TimeEntry, sqlx::Error> {
        sqlx::query(
            "UPDATE timeEntries SET 
             clientId = COALESCE(?, clientId),
             description = COALESCE(?, description),
             startTime = COALESCE(?, startTime),
             endTime = COALESCE(?, endTime),
             duration = COALESCE(?, duration),
             tags = COALESCE(?, tags),
             isBillable = COALESCE(?, isBillable),
             hourlyRate = COALESCE(?, hourlyRate)
             WHERE id = ?"
        )
        .bind(&updates.client_id)
        .bind(&updates.description)
        .bind(&updates.start_time)
        .bind(&updates.end_time)
        .bind(&updates.duration)
        .bind(&updates.tags)
        .bind(&updates.is_billable)
        .bind(&updates.hourly_rate)
        .bind(time_entry_id)
        .execute(&self.pool)
        .await?;

        self.get_time_entry(time_entry_id).await?
            .ok_or_else(|| sqlx::Error::RowNotFound)
    }

    pub async fn delete_time_entry(&self, time_entry_id: &str) -> Result<bool, sqlx::Error> {
        let result = sqlx::query("DELETE FROM timeEntries WHERE id = ?")
            .bind(time_entry_id)
            .execute(&self.pool)
            .await?;

        Ok(result.rows_affected() > 0)
    }
}